### PR TITLE
Verbs fix: Removed Dependency on Duplicate Recv Flag

### PR DESCRIPTION
**Pull request by yanivbl6**
This PR is a response for #11825 

I've added a new Queue and Table for the **RdmaTensorBuffer**, where the relevant information from the send-recv waiter table is stored in case the actual rdma-write does not occur (If the allocated buffer is too small, or busy). When we get a RDMA_MESSAGE_BUFFER_RESPONSE, or RDMA_MESSAGE_BUFFER_IDLE, we check the new queue/table instead of checking the waiter table. (This happens in the new ReSendNextItem() function).

Also added a new function that generates the callback lambda function, to avoid code redundancy.
The result is that the code is less clean- I am open to suggestions on making it more tidy.

@junshi15 @shamoya

### CODE REVIEW

**Comment by shamoya** [comment]
+ @poxvoculi

File: tensorflow/contrib/verbs/rdma.cc [hunk 1]
```diff
@@ -793,9 +841,16 @@ void RdmaTensorBuffer::PostCopyOperations(
       VLOG(2) << "Extend RDMA buffer from " << size_ << " to " << buffer_size;
     }
     CreateCPUBuffer(buffer_size, false);
+    //Need to be recieved again, put into the re-recv queue and the table
+    requeue.push(key_with_step_id);
+    ReItem* item = new ReItem();
+    item->send_args = send_args;
+    item->recv_args = recv_args;
+    item->in = in;
```
Comment (junshi15) [inline tensorflow/contrib/verbs/rdma.cc]: Do we need to increase reference count on [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/rendezvous.cc#L217-L219) or [that](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/rendezvous.cc#L175-L177)?
```
      if (item->recv_args.device_context) {
        item->recv_args.device_context->Ref();
      }
```

**Comment by junshi15** [comment]
@yanivbl6, thanks for the quick response. It looks mostly good. I have a comment regarding the reference count.

**Comment by yanivbl6** [comment]
Thanks @junshi15,
I've added a constructor / destructor to the ReItem struct, to take care of the referencing and the de-referencing.  Also removed the unused field (Status).

**Comment by junshi15** [comment]
It looks good to me. @shamoya any comment?

**Comment by yanivbl6** [comment]
@shamoya participated in the code writing, and reviewed it prior to the PR.

File: tensorflow/contrib/verbs/rdma.h [hunk 2]
```diff
@@ -224,14 +225,52 @@ class RdmaMessageBuffer : public RdmaBuffer {
...
+      const string& key_with_step_id, const string& key,
+      int64 step_id, const Rendezvous::ParsedKey& parsed);
+
+  struct ReItem {
+    Rendezvous::Args send_args;
+    Rendezvous::Args recv_args;
+    Tensor in;
+    bool is_dead;
+
+    ReItem(const Rendezvous::Args& send_args_, const Rendezvous::Args& recv_args_, const Tensor& in_, bool is_dead_ ): send_args(send_args_), recv_args(recv_args_), in(in_), is_dead(is_dead_){
```
Comment (poxvoculi) [inline tensorflow/contrib/verbs/rdma.h]: too long line, break by col 80

File: tensorflow/contrib/verbs/rdma.cc [hunk 3]
```diff
@@ -793,9 +840,12 @@ void RdmaTensorBuffer::PostCopyOperations(
       VLOG(2) << "Extend RDMA buffer from " << size_ << " to " << buffer_size;
     }
     CreateCPUBuffer(buffer_size, false);
+    //Need to be recieved again, put into the re-recv queue and the table
```
Comment (poxvoculi) [inline tensorflow/contrib/verbs/rdma.cc]: s/recieved/received/

**Review by poxvoculi — APPROVED** [review]

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please

**Pull request merged**